mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -214,215 +217,25 @@ final class SemanticTokenProvider(
     val candidates = nodesIterator.dropWhile(_.pos.start < tk.pos.start)
     candidates
       .takeWhile(_.pos.start == tk.pos.start)
-      .find(isTarget) match {
-      case None => None
-      case Some(value) => Some((value, candidates))
-    }
-  }
-
-  case class NodeInfo(
-      sym: Option[Symbol],
-      pos: scala.reflect.api.Position
-  )
-  object NodeInfo {
-    def apply(tree: Tree, pos: scala.reflect.api.Position): NodeInfo = {
-      val sym = tree.symbol
-      if (sym != NoSymbol && sym != null)
-        NodeInfo(Some(tree.symbol), pos)
-      else NodeInfo(None, pos)
-    }
-
-    def apply(sym: Symbol, pos: scala.reflect.api.Position): NodeInfo =
-      new NodeInfo(Some(sym), pos)
-  }
-
-  /**
-   * was written in reference to PcDocumentHighlightProvider.
-   */
-  import cp._
-  object traverser {
-
-    /**
-     * gathers all nodes inside given tree.
-     * The nodes have symbol.
-     */
-    def traverse(
-        nodes: SortedSet[NodeInfo],
-        tree: cp.Tree
-    ): SortedSet[NodeInfo] = {
-
-      tree match {
-        /**
-         * All indentifiers such as:
-         * val a = <<b>>
-         */
-        case ident: cp.Ident if ident.pos.isRange =>
-          val symbol =
-            if (ident.symbol == NoSymbol)
-              if (ident.tpe != null) {
-                ident.tpe.underlying.typeSymbolDirect
-              } else {
-                val context = doLocateContext(ident.pos)
-                context.lookupSymbol(ident.name, _ => true) match {
-                  case LookupSucceeded(_, symbol) => symbol
-                  case _ => NoSymbol
-                }
-              }
-            else {
-              ident.symbol
+      .filter(isTarget) match {
+      case Nil => None
+      case node :: Nil => Some((node, candidates))
+      case manyNodes =>
+        manyNodes.filter(ni =>
+          Collector
+            .symbolAlternatives(ni.sym)
+            .exists(_.decodedName == tk.text)
+        ) match {
+          case Nil => None
+          case matchingNames
```

Comment:
In case there are multiple symbols at the same position, we first filter out those with not matching name.
If there is still more than one symbol, we handle the case for class declaration
```scala
class Foo(a: Int, b: Int)
```
On `a` and `b` are two symbols: parameter in constructor method and `val`, we want to choose `val`.